docs/P4-Compilation.md

```diff
@@ -0,0 +1,131 @@
+Compiling your own P4 code
+==========================
+
+This document explains how to compile your own P4 code, that can be located anywhere outside the open-P4studio tree.
+
+## A little theory...
+
+On the surface, the P4 compiler for Tofino (`bf-p4c`) works like any other compiler. In other words, it is totally ok to compile your program using the following command line:
+
+```
+bf-p4c my_program.p4
+```
+
+This should work for any simple program, written for Tofino. At the end of the compilation process the compiler will create the directory `my_program.tofino`, containing all compilation artifacts. If your program is intended for Tofino2, add the parameter `--target tofino2`; the name of the compilation directory will be `my_program.tofino2` then.
+
+In more complex cases, you might need to provide addiional parameters, similar to the ones used by most C compilers, such as `-I <include_dir>`, `-DCPP_VAR[=value]` or `-UCPP_VAR`. It is also highly recommended to add the `-g` parameter, so that the compiler can output the detailed logs and placement information (especially if you have access to the `p4i` (P4Insight) tool). 
+
+The compiler has many other special options and they might sometimse be required, but they are not the focus of this document.
+
+While the command above is the easiest way to invoke the compiler and is often used by the developers to check whether the program compiles or not, it is rarely used. The reason is that after compiling a program we need to exercise it by starting the user-space driver and loading the program onto the target, be it either a Tofino model or the real ASIC. In the simulation case it is also important to allow the model to load the program layout information so that it can produce human-readable logs by translating low-level operations back into the familiar symbols from your program. 
+
+Generally speaking, the compiler produces three main artifacts that are neccessary to run a P4 program:
+
+1. The compiled program binary, called `tofino.bin` (or `tofino2.bin`) that needs to be loaded into the ASIC or the model so that they can execute the P4 code
+2. The `bf-rt.json` file that describes all the objects, defined in the P4 program, such as tables and their key fields, actions and their action data fields, parser value sets and externs
```

Comment:
The file I saw created was `bfrt.json`, not `bf-rt.json`

Reply:
When I followed the instructions using `cmake` below, I did see a file named `bf-rt.json` created.

Reply:
This is funny! I haven't noticed that previously, most probably because as explained in this section that directory that is created when running the compiler directly is never used to run the program. 

As I would always say in those cases "the great artists never repeat themselves" and thus P4 Studio code is full of inconsistencies. This one can probably fixed pretty easily if we decide to do that.

For the sake of the doc I can say "`bf_rt.json` (sometimes also called `bfrt.json`)". Or vice versa. Will that help?